// @syncfusion/ej2-pdfviewer/src/pdfviewer/toolbar/toolbar.js
import { createElement } from '@syncfusion/ej2-base';
import { Toolbar as tool } from '@syncfusion/ej2-navigations';
import { NumericTextBox } from '@syncfusion/ej2-inputs';
import { ComboBox } from '@syncfusion/ej2-dropdowns';
import { Tooltip } from '@syncfusion/ej2-popups';
/**
 * Toolbar module
 */
var Toolbar = /** @class */ (function () {
    /**
     * @private
     */
    function Toolbar(viewer, viewerBase) {
        var _this = this;
        this.isPageNavigationToolDisabled = false;
        this.isMagnificationToolDisabled = false;
        this.isSelectionToolDisabled = false;
        this.isScrollingToolDisabled = false;
        this.isOpenBtnVisible = true;
        this.isNavigationToolVisible = true;
        this.isMagnificationToolVisible = true;
        this.isSelectionBtnVisible = true;
        this.isScrollingBtnVisible = true;
        this.isDownloadBtnVisible = true;
        this.isPrintBtnVisible = true;
        this.isSearchBtnVisible = true;
        this.isTextSearchBoxDisplayed = false;
        this.onToolbarKeydown = function (event) {
            var targetId = event.target.id;
            if (!(targetId === _this.pdfViewer.element.id + '_currentPageInput' || targetId === _this.pdfViewer.element.id + '_zoomDropDown')) {
                event.preventDefault();
                event.stopPropagation();
            }
        };
        this.toolbarClickHandler = function (args) {
            // tslint:disable-next-line:max-line-length
            if (args.originalEvent.target === _this.zoomDropdownItem.parentElement.childNodes[1] || args.originalEvent.target === _this.zoomDropdownItem.parentElement.childNodes[2]) {
                args.cancel = true;
            }
            else if (args.originalEvent.target.id === _this.pdfViewer.element.id + '_openIcon') {
                var tooltipData = args.originalEvent.target.parentElement.dataset;
                if (tooltipData && tooltipData.tooltipId) {
                    var tooltipElement = document.getElementById(tooltipData.tooltipId);
                    if (tooltipElement) {
                        tooltipElement.style.display = 'none';
                    }
                }
            }
            switch (args.originalEvent.target.id) {
                case _this.pdfViewer.element.id + '_open':
                case _this.pdfViewer.element.id + '_openIcon':
                    _this.fileInputElement.click();
                    break;
                case _this.pdfViewer.element.id + '_download':
                case _this.pdfViewer.element.id + '_downloadIcon':
                    _this.pdfViewerBase.download();
                    break;
                case _this.pdfViewer.element.id + '_print':
                case _this.pdfViewer.element.id + '_printIcon':
                    if (_this.pdfViewer.printModule) {
                        _this.pdfViewer.printModule.print();
                    }
                    break;
                case _this.pdfViewer.element.id + '_firstPage':
                case _this.pdfViewer.element.id + '_firstPageIcon':
                    if (_this.pdfViewer.navigationModule) {
                        _this.pdfViewer.navigationModule.goToFirstPage();
                    }
                    break;
                case _this.pdfViewer.element.id + '_previousPage':
                case _this.pdfViewer.element.id + '_previousPageIcon':
                    if (_this.pdfViewer.navigationModule) {
                        _this.pdfViewer.navigationModule.goToPreviousPage();
                    }
                    break;
                case _this.pdfViewer.element.id + '_nextPage':
                case _this.pdfViewer.element.id + '_nextPageIcon':
                    if (_this.pdfViewer.navigationModule) {
                        _this.pdfViewer.navigationModule.goToNextPage();
                    }
                    break;
                case _this.pdfViewer.element.id + '_lastPage':
                case _this.pdfViewer.element.id + '_lastPageIcon':
                    if (_this.pdfViewer.navigationModule) {
                        _this.pdfViewer.navigationModule.goToLastPage();
                    }
                    break;
                case _this.pdfViewer.element.id + '_zoomIn':
                case _this.pdfViewer.element.id + '_zoomInIcon':
                    _this.pdfViewer.magnificationModule.zoomIn();
                    break;
                case _this.pdfViewer.element.id + '_zoomOut':
                case _this.pdfViewer.element.id + '_zoomOutIcon':
                    _this.pdfViewer.magnificationModule.zoomOut();
                    break;
                case _this.pdfViewer.element.id + '_selectTool':
                case _this.pdfViewer.element.id + '_selectToolIcon':
                    if (!_this.isSelectionToolDisabled) {
                        _this.pdfViewerBase.initiateTextSelectMode();
                        _this.updateInteractionTools(true);
                    }
                    break;
                case _this.pdfViewer.element.id + '_handTool':
                case _this.pdfViewer.element.id + '_handToolIcon':
                    if (!_this.isScrollingToolDisabled) {
                        _this.pdfViewerBase.initiatePanning();
                        _this.updateInteractionTools(false);
                    }
                    break;
                case _this.pdfViewer.element.id + '_search':
                case _this.pdfViewer.element.id + '_searchIcon':
                    _this.textSearchButtonHandler();
                    break;
            }
            // tslint:disable-next-line:max-line-length
            if (!(args.originalEvent.target === _this.zoomDropdownItem.parentElement.childNodes[1] || args.originalEvent.target === _this.zoomDropdownItem.parentElement.childNodes[2] || args.originalEvent.target === _this.currentPageBoxElement || args.originalEvent.target === _this.textSearchItem.childNodes[0])) {
                args.originalEvent.target.blur();
                _this.pdfViewerBase.focusViewerContainer();
            }
        };
        // tslint:disable-next-line
        this.loadDocument = function (args) {
            // tslint:disable-next-line
            var upoadedFiles = args.target.files;
            if (args.target.files[0] !== null) {
                var uploadedFile = upoadedFiles[0];
                if (uploadedFile) {
                    _this.uploadedDocumentName = uploadedFile.name;
                    var reader = new FileReader();
                    reader.readAsDataURL(uploadedFile);
                    // tslint:disable-next-line
                    reader.onload = function (e) {
                        var uploadedFileUrl = e.currentTarget.result;
                        _this.pdfViewer.load(uploadedFileUrl, null);
                    };
                }
            }
        };
        this.navigateToPage = function (args) {
            if (args.which === 13) {
                // tslint:disable-next-line
                var enteredValue = parseInt(_this.currentPageBoxElement.value);
                if (enteredValue !== null) {
                    if (enteredValue > 0 && enteredValue <= _this.pdfViewerBase.pageCount) {
                        if (_this.pdfViewer.navigationModule) {
                            _this.pdfViewer.navigationModule.goToPage(enteredValue);
                        }
                    }
                    else {
                        _this.updateCurrentPage(_this.pdfViewerBase.currentPageNumber);
                    }
                }
                else {
                    _this.updateCurrentPage(_this.pdfViewerBase.currentPageNumber);
                }
                _this.currentPageBoxElement.blur();
                _this.pdfViewerBase.focusViewerContainer();
            }
        };
        this.textBoxFocusOut = function () {
            // tslint:disable-next-line
            if (_this.currentPageBox.value === null || _this.currentPageBox.value >= _this.pdfViewerBase.pageCount || _this.currentPageBox.value !== _this.pdfViewerBase.currentPageNumber) {
                _this.updateCurrentPage(_this.pdfViewerBase.currentPageNumber);
            }
        };
        this.pdfViewer = viewer;
        this.pdfViewerBase = viewerBase;
    }
    /**
     * @private
     */
    Toolbar.prototype.intializeToolbar = function (width) {
        var toolbarDiv = this.createToolbar(width);
        // tslint:disable-next-line
        var isIE = !!document.documentMode;
        if (isIE) {
            this.totalPageItem.classList.add('e-pv-total-page-ms');
        }
        this.createFileElement(toolbarDiv);
        this.wireEvent();
        this.updateToolbarItems();
        this.applyToolbarSettings();
        this.initialEnableItems();
        return toolbarDiv;
    };
    /**
     * Shows /hides the toolbar in the PdfViewer
     * @param  {boolean} enableToolbar
     * @returns void
     */
    Toolbar.prototype.showToolbar = function (enableToolbar) {
        var toolbar = this.toolbarElement;
        if (enableToolbar) {
            toolbar.style.display = 'block';
        }
        else {
            toolbar.style.display = 'none';
            this.pdfViewerBase.navigationPane.sideBarToolbar.style.display = 'none';
        }
    };
    /**
     * Shows /hides the the toolbar items in the PdfViewer
     * @param  {string[]} items
     * @param  {boolean} isVisible
     * @returns void
     */
    Toolbar.prototype.showToolbarItem = function (items, isVisible) {
        for (var i = 0; i < items.length; i++) {
            switch (items[i]) {
                case 'OpenOption':
                    this.showOpenOption(isVisible);
                    break;
                case 'PageNavigationTool':
                    this.showPageNavigationTool(isVisible);
                    break;
                case 'MagnificationTool':
                    this.showMagnificationTool(isVisible);
                    break;
                case 'SelectionTool':
                    this.showSelectionTool(isVisible);
                    break;
                case 'PanTool':
                    this.showScrollingTool(isVisible);
                    break;
                case 'DownloadOption':
                    this.showDownloadOption(isVisible);
                    break;
                case 'PrintOption':
                    this.showPrintOption(isVisible);
                    break;
                case 'SearchOption':
                    this.showSearchOption(isVisible);
                    break;
            }
        }
        this.applyHideToToolbar(true, 1, 1);
        this.applyHideToToolbar(true, 8, 8);
        this.applyHideToToolbar(true, 12, 12);
        this.showSeparator(items);
    };
    /**
     * Enables /disables the the toolbar items in the PdfViewer
     * @param  {string[]} items
     * @param  {boolean} isEnable
     * @returns void
     */
    Toolbar.prototype.enableToolbarItem = function (items, isEnable) {
        for (var i = 0; i < items.length; i++) {
            switch (items[i]) {
                case 'OpenOption':
                    this.enableOpenOption(isEnable);
                    break;
                case 'PageNavigationTool':
                    this.isPageNavigationToolDisabled = isEnable;
                    this.enablePageNavigationTool(isEnable);
                    break;
                case 'MagnificationTool':
                    this.isMagnificationToolDisabled = isEnable;
                    this.enableMagnificationTool(isEnable);
                    break;
                case 'SelectionTool':
                    this.isSelectionToolDisabled = isEnable;
                    this.enableSelectionTool(isEnable);
                    break;
                case 'PanTool':
                    this.isScrollingToolDisabled = isEnable;
                    this.enableScrollingTool(isEnable);
                    break;
                case 'DownloadOption':
                    this.enableDownloadOption(isEnable);
                    break;
                case 'PrintOption':
                    this.enablePrintOption(isEnable);
                    break;
                case 'SearchOption':
                    this.enableSearchOption(isEnable);
                    break;
            }
        }
    };
    Toolbar.prototype.showOpenOption = function (enableOpenOption) {
        this.isOpenBtnVisible = enableOpenOption;
        this.applyHideToToolbar(enableOpenOption, 0, 0);
    };
    Toolbar.prototype.showPageNavigationTool = function (enablePageNavigationTool) {
        this.isNavigationToolVisible = enablePageNavigationTool;
        this.applyHideToToolbar(enablePageNavigationTool, 2, 7);
    };
    Toolbar.prototype.showMagnificationTool = function (enableMagnificationTool) {
        this.isMagnificationToolVisible = enableMagnificationTool;
        this.applyHideToToolbar(enableMagnificationTool, 9, 11);
    };
    Toolbar.prototype.showSelectionTool = function (enableSelectionTool) {
        this.isSelectionBtnVisible = enableSelectionTool;
        this.applyHideToToolbar(enableSelectionTool, 13, 13);
    };
    Toolbar.prototype.showScrollingTool = function (enableScrollingTool) {
        this.isScrollingBtnVisible = enableScrollingTool;
        this.applyHideToToolbar(enableScrollingTool, 14, 14);
    };
    Toolbar.prototype.showDownloadOption = function (enableDownloadOption) {
        this.isDownloadBtnVisible = enableDownloadOption;
        this.applyHideToToolbar(enableDownloadOption, 16, 16);
    };
    Toolbar.prototype.showPrintOption = function (enablePrintOption) {
        this.isPrintBtnVisible = enablePrintOption;
        this.applyHideToToolbar(enablePrintOption, 17, 17);
    };
    Toolbar.prototype.showSearchOption = function (enableSearchOption) {
        this.isSearchBtnVisible = enableSearchOption;
        this.applyHideToToolbar(enableSearchOption, 15, 15);
    };
    Toolbar.prototype.enableOpenOption = function (enableOpenOption) {
        this.toolbar.enableItems(this.openDocumentItem.parentElement, enableOpenOption);
    };
    Toolbar.prototype.enablePageNavigationTool = function (enablePageNavigationTool) {
        this.toolbar.enableItems(this.firstPageItem.parentElement, enablePageNavigationTool);
        this.toolbar.enableItems(this.previousPageItem.parentElement, enablePageNavigationTool);
        this.toolbar.enableItems(this.nextPageItem.parentElement, enablePageNavigationTool);
        this.toolbar.enableItems(this.lastPageItem.parentElement, enablePageNavigationTool);
        this.currentPageBox.readonly = !enablePageNavigationTool;
    };
    Toolbar.prototype.enableMagnificationTool = function (enableMagnificationTool) {
        this.toolbar.enableItems(this.zoomInItem.parentElement, enableMagnificationTool);
        this.toolbar.enableItems(this.zoomOutItem.parentElement, enableMagnificationTool);
        this.zoomDropDown.readonly = !enableMagnificationTool;
    };
    Toolbar.prototype.enableSelectionTool = function (enableSelectionTool) {
        this.toolbar.enableItems(this.textSelectItem.parentElement, enableSelectionTool);
    };
    Toolbar.prototype.enableScrollingTool = function (enableScrollingTool) {
        this.toolbar.enableItems(this.panItem.parentElement, enableScrollingTool);
    };
    Toolbar.prototype.enableDownloadOption = function (enableDownloadOption) {
        this.toolbar.enableItems(this.downloadItem.parentElement, enableDownloadOption);
    };
    Toolbar.prototype.enablePrintOption = function (enablePrintOption) {
        this.toolbar.enableItems(this.printItem.parentElement, enablePrintOption);
    };
    Toolbar.prototype.enableSearchOption = function (enableSearchOption) {
        this.toolbar.enableItems(this.textSearchItem.parentElement, enableSearchOption);
    };
    /**
     * @private
     */
    Toolbar.prototype.resetToolbar = function () {
        this.currentPageBox.min = 0;
        this.currentPageBox.value = 0;
        this.updateTotalPage();
        this.updateToolbarItems();
    };
    /**
     * @private
     */
    Toolbar.prototype.updateToolbarItems = function () {
        if (this.pdfViewerBase.pageCount === 0) {
            this.toolbar.enableItems(this.downloadItem.parentElement, false);
            this.toolbar.enableItems(this.printItem.parentElement, false);
            this.updateNavigationButtons();
            this.toolbar.enableItems(this.zoomInItem.parentElement, false);
            this.toolbar.enableItems(this.zoomOutItem.parentElement, false);
            if (this.pdfViewer.magnificationModule) {
                this.zoomDropDown.readonly = true;
            }
            this.toolbar.enableItems(this.pdfViewerBase.getElement('_currentPageInputContainer'), false);
            this.toolbar.enableItems(this.pdfViewerBase.getElement('_zoomDropDownContainer'), false);
            this.toolbar.enableItems(this.textSelectItem.parentElement, false);
            this.toolbar.enableItems(this.panItem.parentElement, false);
            this.toolbar.enableItems(this.textSearchItem.parentElement, false);
        }
        else if (this.pdfViewerBase.pageCount > 0) {
            this.toolbar.enableItems(this.downloadItem.parentElement, true);
            this.toolbar.enableItems(this.printItem.parentElement, true);
            this.toolbar.enableItems(this.pdfViewerBase.getElement('_currentPageInputContainer'), true);
            this.toolbar.enableItems(this.pdfViewerBase.getElement('_zoomDropDownContainer'), true);
            this.updateNavigationButtons();
            this.updateZoomButtons();
            if (this.pdfViewer.magnificationModule) {
                this.zoomDropDown.readonly = false;
            }
            this.updateInteractionItems();
            if (this.pdfViewer.textSearchModule && this.pdfViewer.enableTextSearch) {
                this.toolbar.enableItems(this.textSearchItem.parentElement, true);
            }
        }
    };
    /**
     * @private
     */
    Toolbar.prototype.updateNavigationButtons = function () {
        if (this.pdfViewer.navigationModule && !this.isPageNavigationToolDisabled) {
            if (this.pdfViewerBase.pageCount === 0 || (this.pdfViewerBase.currentPageNumber === 1 && this.pdfViewerBase.pageCount === 1)) {
                this.toolbar.enableItems(this.firstPageItem.parentElement, false);
                this.toolbar.enableItems(this.previousPageItem.parentElement, false);
                this.toolbar.enableItems(this.nextPageItem.parentElement, false);
                this.toolbar.enableItems(this.lastPageItem.parentElement, false);
            }
            else if (this.pdfViewerBase.currentPageNumber === 1 && this.pdfViewerBase.pageCount > 0) {
                this.toolbar.enableItems(this.firstPageItem.parentElement, false);
                this.toolbar.enableItems(this.previousPageItem.parentElement, false);
                this.toolbar.enableItems(this.nextPageItem.parentElement, true);
                this.toolbar.enableItems(this.lastPageItem.parentElement, true);
            }
            else if (this.pdfViewerBase.currentPageNumber === this.pdfViewerBase.pageCount && this.pdfViewerBase.pageCount > 0) {
                this.toolbar.enableItems(this.firstPageItem.parentElement, true);
                this.toolbar.enableItems(this.previousPageItem.parentElement, true);
                this.toolbar.enableItems(this.nextPageItem.parentElement, false);
                this.toolbar.enableItems(this.lastPageItem.parentElement, false);
            }
            else if (this.pdfViewerBase.currentPageNumber > 1 && this.pdfViewerBase.currentPageNumber < this.pdfViewerBase.pageCount) {
                this.toolbar.enableItems(this.firstPageItem.parentElement, true);
                this.toolbar.enableItems(this.previousPageItem.parentElement, true);
                this.toolbar.enableItems(this.nextPageItem.parentElement, true);
                this.toolbar.enableItems(this.lastPageItem.parentElement, true);
            }
        }
        else {
            this.toolbar.enableItems(this.firstPageItem.parentElement, false);
            this.toolbar.enableItems(this.previousPageItem.parentElement, false);
            this.toolbar.enableItems(this.nextPageItem.parentElement, false);
            this.toolbar.enableItems(this.lastPageItem.parentElement, false);
            this.currentPageBox.readonly = true;
        }
    };
    /**
     * @private
     */
    Toolbar.prototype.updateZoomButtons = function () {
        if (this.pdfViewer.magnificationModule && !this.isMagnificationToolDisabled) {
            if (this.pdfViewer.magnificationModule.zoomFactor <= 0.5) {
                this.toolbar.enableItems(this.zoomInItem.parentElement, true);
                this.toolbar.enableItems(this.zoomOutItem.parentElement, false);
            }
            else if (this.pdfViewer.magnificationModule.zoomFactor >= 4) {
                this.toolbar.enableItems(this.zoomInItem.parentElement, false);
                this.toolbar.enableItems(this.zoomOutItem.parentElement, true);
            }
            else {
                this.toolbar.enableItems(this.zoomInItem.parentElement, true);
                this.toolbar.enableItems(this.zoomOutItem.parentElement, true);
            }
        }
    };
    /**
     * @private
     */
    Toolbar.prototype.destroy = function () {
        this.unWireEvent();
        this.toolbar.destroy();
        this.toolbarElement.remove();
    };
    /**
     * @private
     */
    Toolbar.prototype.updateCurrentPage = function (pageIndex) {
        if (this.currentPageBox.value === pageIndex) {
            this.currentPageBoxElement.value = pageIndex.toString();
        }
        this.currentPageBox.value = pageIndex;
        this.pdfViewerBase.currentPageNumber = pageIndex;
    };
    /**
     * @private
     */
    Toolbar.prototype.updateTotalPage = function () {
        if (this.pdfViewerBase.pageCount > 0) {
            this.currentPageBox.min = 1;
        }
        this.totalPageItem.textContent = 'of ' + this.pdfViewerBase.pageCount.toString();
    };
    /**
     * @private
     */
    Toolbar.prototype.openFileDialogBox = function (event) {
        event.preventDefault();
        this.fileInputElement.click();
    };
    Toolbar.prototype.createToolbar = function (controlWidth) {
        // tslint:disable-next-line:max-line-length
        this.toolbarElement = createElement('div', { id: this.pdfViewer.element.id + '_toolbarContainer', className: 'e-pv-toolbar' });
        this.pdfViewerBase.viewerMainContainer.appendChild(this.toolbarElement);
        this.itemsContainer = createElement('div', { id: this.pdfViewer.element.id + '_toolbarItemsContainer' });
        // tslint:disable-next-line:max-line-length
        var openButtonContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_open', this.pdfViewer.localeObj.getConstant('Open'), 'e-pv-open-document');
        this.openDocumentItem = openButtonContainer.firstChild;
        this.itemsContainer.appendChild(openButtonContainer);
        this.toolbarElement.appendChild(this.itemsContainer);
        var seperatorDiv1 = createElement('div', { className: 'e-separator' });
        this.itemsContainer.appendChild(seperatorDiv1);
        // tslint:disable-next-line:max-line-length
        var firstPageContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_firstPage', this.pdfViewer.localeObj.getConstant('Go To First Page'), 'e-pv-first-page-navigation');
        this.firstPageItem = firstPageContainer.firstChild;
        this.itemsContainer.appendChild(firstPageContainer);
        // tslint:disable-next-line:max-line-length
        var previousContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_previousPage', this.pdfViewer.localeObj.getConstant('Previous Page'), 'e-pv-previous-page-navigation');
        this.previousPageItem = previousContainer.firstChild;
        this.itemsContainer.appendChild(previousContainer);
        // tslint:disable-next-line:max-line-length
        var nextContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_nextPage', this.pdfViewer.localeObj.getConstant('Next Page'), 'e-pv-next-page-navigation');
        this.nextPageItem = nextContainer.firstChild;
        this.itemsContainer.appendChild(nextContainer);
        // tslint:disable-next-line:max-line-length
        var lastPageContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_lastPage', this.pdfViewer.localeObj.getConstant('Go To Last Page'), 'e-pv-last-page-navigation');
        this.lastPageItem = lastPageContainer.firstChild;
        this.itemsContainer.appendChild(lastPageContainer);
        // tslint:disable-next-line:max-line-length
        var goToPageContainer = this.createToolbarItem('input', this.pdfViewer.element.id + '_currentPageInput', this.pdfViewer.localeObj.getConstant('Page Number'), null);
        this.itemsContainer.appendChild(goToPageContainer);
        this.currentPageBox = new NumericTextBox({ value: 0, format: '##', cssClass: 'e-pv-current-page-box', showSpinButton: false });
        this.currentPageBoxElement = goToPageContainer.firstChild;
        this.currentPageBox.appendTo(this.currentPageBoxElement);
        // tslint:disable-next-line:max-line-length
        var totalPageContainer = this.createToolbarItem('span', this.pdfViewer.element.id + '_totalPage', null, 'e-pv-total-page');
        this.totalPageItem = totalPageContainer.firstChild;
        this.itemsContainer.appendChild(totalPageContainer);
        this.updateTotalPage();
        var seperatorDiv2 = createElement('div', { className: 'e-separator' });
        this.itemsContainer.appendChild(seperatorDiv2);
        // tslint:disable-next-line:max-line-length
        var zoomOutContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_zoomOut', this.pdfViewer.localeObj.getConstant('Zoom Out'), 'e-pv-zoom-out');
        this.zoomOutItem = zoomOutContainer.firstChild;
        this.itemsContainer.appendChild(zoomOutContainer);
        // tslint:disable-next-line:max-line-length
        var zoomInContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_zoomIn', this.pdfViewer.localeObj.getConstant('Zoom In'), 'e-pv-zoom-in');
        this.zoomInItem = zoomInContainer.firstChild;
        this.itemsContainer.appendChild(zoomInContainer);
        // tslint:disable-next-line:max-line-length
        var zoomDropdownContainer = this.createToolbarItem('input', this.pdfViewer.element.id + '_zoomDropDown', this.pdfViewer.localeObj.getConstant('Zoom'), null);
        zoomDropdownContainer.className = zoomDropdownContainer.className + ' e-pv-zoom-drop-down-container';
        this.zoomDropdownItem = zoomDropdownContainer.firstChild;
        this.itemsContainer.appendChild(zoomDropdownContainer);
        var seperatorDiv3 = createElement('div', { className: 'e-separator' });
        this.itemsContainer.appendChild(seperatorDiv3);
        // tslint:disable-next-line:max-line-length
        var selectToolContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_selectTool', this.pdfViewer.localeObj.getConstant('Text Selection'), 'e-pv-text-select-tool');
        this.textSelectItem = selectToolContainer.firstChild;
        this.itemsContainer.appendChild(selectToolContainer);
        // tslint:disable-next-line:max-line-length
        var handToolContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_handTool', this.pdfViewer.localeObj.getConstant('Panning'), 'e-pv-pan-tool');
        this.panItem = handToolContainer.firstChild;
        this.itemsContainer.appendChild(handToolContainer);
        // tslint:disable-next-line:max-line-length
        var searchButtonContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_search', this.pdfViewer.localeObj.getConstant('Text Search'), 'e-pv-text-search');
        this.textSearchItem = searchButtonContainer.firstChild;
        this.itemsContainer.appendChild(searchButtonContainer);
        searchButtonContainer.style.position = 'absolute';
        // tslint:disable-next-line:max-line-length
        var downloadButtonContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_download', this.pdfViewer.localeObj.getConstant('Download'), 'e-pv-download-document');
        this.downloadItem = downloadButtonContainer.firstChild;
        downloadButtonContainer.style.position = 'absolute';
        // tslint:disable-next-line:max-line-length
        downloadButtonContainer.style.left = (this.pdfViewer.element.clientWidth - this.pdfViewerBase.navigationPane.getViewerContainerLeft()) + 'px';
        this.itemsContainer.appendChild(downloadButtonContainer);
        // tslint:disable-next-line:max-line-length
        var printButtonContainer = this.createToolbarItem('button', this.pdfViewer.element.id + '_print', this.pdfViewer.localeObj.getConstant('Print'), 'e-pv-print-document');
        this.printItem = printButtonContainer.firstChild;
        printButtonContainer.style.position = 'absolute';
        if (downloadButtonContainer !== undefined) { // tslint:disable-next-line:max-line-length
            printButtonContainer.style.left = (this.pdfViewer.element.clientWidth - this.pdfViewerBase.navigationPane.getViewerContainerLeft() - 43) + 'px';
        }
        else { // tslint:disable-next-line:max-line-length
            printButtonContainer.style.left = (this.pdfViewer.element.clientWidth - this.pdfViewerBase.navigationPane.getViewerContainerLeft()) + 'px';
        } // tslint:disable-next-line:max-line-length
        searchButtonContainer.style.left = (this.pdfViewer.element.clientWidth - this.pdfViewerBase.navigationPane.getViewerContainerLeft() - downloadButtonContainer.clientWidth - printButtonContainer.clientWidth - 54) + 'px';
        this.itemsContainer.appendChild(printButtonContainer);
        // tslint:disable-next-line:max-line-length
        var items = [{ percent: '50%', id: '0' }, { percent: '75%', id: '1' }, { percent: '100%', id: '2' }, { percent: '125%', id: '3' },
            // tslint:disable-next-line:max-line-length
            { percent: '150%', id: '4' }, { percent: '200%', id: '5' }, { percent: '400%', id: '6' }, { percent: this.pdfViewer.localeObj.getConstant('Fit Page'), id: '7' }, { percent: this.pdfViewer.localeObj.getConstant('Fit Width'), id: '8' }, { percent: this.pdfViewer.localeObj.getConstant('Automatic'), id: '9' }
        ];
        // tslint:disable-next-line:max-line-length
        this.zoomDropDown = new ComboBox({ dataSource: items, text: '100%', fields: { text: 'percent', value: 'id' }, readonly: true, cssClass: 'e-pv-zoom-drop-down', popupHeight: '402px', showClearButton: false });
        this.zoomDropDown.appendTo(this.zoomDropdownItem);
        this.toolbar = new tool({ clicked: this.toolbarClickHandler, width: '', height: '', overflowMode: 'Popup' });
        this.toolbar.appendTo(this.toolbarElement);
        this.toolbarElement.addEventListener('keydown', this.onToolbarKeydown);
        return this.toolbarElement;
    };
    Toolbar.prototype.createToolbarItem = function (elementName, id, tooltipText, className) {
        var containerElement = createElement('div', { id: id + 'Container' });
        var toolbarItem = createElement(elementName, { id: id });
        if (className !== null) {
            containerElement.className = className + '-container e-overflow-show e-popup-text';
            toolbarItem.className = className;
        }
        if (elementName === 'button' && id !== this.pdfViewer.element.id + '_zoomDropDown') {
            toolbarItem.className = 'e-btn e-tbar-btn e-icon-btn e-pv-tbar-btn ' + className;
            var buttonSpan = createElement('span', { id: id + 'Icon', className: className + '-icon e-pv-icon' });
            toolbarItem.appendChild(buttonSpan);
        }
        else if (elementName === 'input' && id !== this.pdfViewer.element.id + '_zoomDropDown') {
            toolbarItem.type = 'text';
        }
        containerElement.appendChild(toolbarItem);
        if (tooltipText !== null) {
            // tslint:disable-next-line
            var tooltip = new Tooltip({ content: tooltipText, opensOn: 'Hover', beforeOpen: this.onTooltipBeforeOpen.bind(this) });
            tooltip.appendTo(toolbarItem);
        }
        return containerElement;
    };
    Toolbar.prototype.onTooltipBeforeOpen = function (args) {
        if (!this.pdfViewer.toolbarSettings.showTooltip) {
            args.cancel = true;
        }
    };
    Toolbar.prototype.createFileElement = function (toolbarElement) {
        // tslint:disable-next-line:max-line-length
        this.fileInputElement = createElement('input', { id: this.pdfViewer.element.id + '_fileUploadElement', styles: 'position:fixed; left:-100em', attrs: { 'type': 'file' } });
        this.fileInputElement.setAttribute('accept', '.pdf');
        toolbarElement.appendChild(this.fileInputElement);
    };
    Toolbar.prototype.wireEvent = function () {
        this.toolbarElement.addEventListener('mouseup', this.toolbarOnMouseup.bind(this));
        this.fileInputElement.addEventListener('change', this.loadDocument);
        this.currentPageBoxElement.addEventListener('focusout', this.textBoxFocusOut);
        this.currentPageBoxElement.addEventListener('keypress', this.navigateToPage);
        this.zoomDropDown.change = this.zoomPercentSelect.bind(this);
        this.zoomDropDown.element.addEventListener('keypress', this.onZoomDropDownInput.bind(this));
        this.zoomDropDown.element.addEventListener('click', this.onZoomDropDownInputClick.bind(this));
    };
    Toolbar.prototype.unWireEvent = function () {
        this.toolbarElement.removeEventListener('mouseup', this.toolbarOnMouseup.bind(this));
        this.fileInputElement.removeEventListener('change', this.loadDocument);
        this.currentPageBoxElement.removeEventListener('focusout', this.textBoxFocusOut);
        this.currentPageBoxElement.removeEventListener('keypress', this.navigateToPage);
        this.zoomDropDown.removeEventListener('change', this.zoomPercentSelect);
        this.zoomDropDown.element.removeEventListener('keypress', this.onZoomDropDownInput);
        this.zoomDropDown.element.removeEventListener('click', this.onZoomDropDownInputClick);
    };
    /**
     * @private
     */
    Toolbar.prototype.onToolbarResize = function (viewerWidth) {
        var navigationToolbar = document.getElementById(this.pdfViewer.element.id + '_toolbarContainer_nav');
        var downloadButtonElement = document.getElementById(this.pdfViewer.element.id + '_downloadContainer');
        var printButtonElement = document.getElementById(this.pdfViewer.element.id + '_printContainer');
        var searchButtonElement = document.getElementById(this.pdfViewer.element.id + '_searchContainer');
        if (!navigationToolbar) {
            if (downloadButtonElement) {
                downloadButtonElement.style.left = viewerWidth + 'px';
                downloadButtonElement.style.position = 'absolute';
            }
            if (printButtonElement) {
                if (downloadButtonElement !== null) {
                    // calculate the print Icon left Position value
                    printButtonElement.style.left = (viewerWidth - downloadButtonElement.clientWidth) + 'px';
                    printButtonElement.style.position = 'absolute';
                }
                else {
                    printButtonElement.style.left = viewerWidth + 'px';
                    printButtonElement.style.position = 'absolute';
                }
            }
            if (searchButtonElement) {
                searchButtonElement.style.position = 'absolute';
                if (printButtonElement !== null) {
                    // calculate the search Icon left Position value
                    searchButtonElement.style.left = (parseFloat(printButtonElement.style.left) - printButtonElement.clientWidth) + 'px';
                }
                else if (downloadButtonElement !== null) {
                    // tslint:disable-next-line:max-line-length
                    searchButtonElement.style.left = (parseFloat(downloadButtonElement.style.left) - downloadButtonElement.clientWidth) + 'px';
                }
                else {
                    searchButtonElement.style.left = viewerWidth + 'px';
                }
            }
        }
    };
    Toolbar.prototype.toolbarOnMouseup = function (event) {
        if (event.target === this.itemsContainer || event.target === this.toolbarElement) {
            this.pdfViewerBase.focusViewerContainer();
        }
    };
    Toolbar.prototype.applyHideToToolbar = function (show, startIndex, endIndex) {
        var isHide = !show;
        for (var index = startIndex; index <= endIndex; index++) {
            this.toolbar.hideItem(index, isHide);
        }
    };
    Toolbar.prototype.onZoomDropDownInput = function (event) {
        if ((event.which < 48 || event.which > 57) && event.which !== 8 && event.which !== 13) {
            event.preventDefault();
            return false;
        }
        else {
            if (event.which === 13) {
                event.preventDefault();
                var value = this.zoomDropDown.element.value;
                this.zoomDropDownChange(value);
            }
            return true;
        }
    };
    Toolbar.prototype.onZoomDropDownInputClick = function () {
        this.zoomDropDown.element.select();
    };
    Toolbar.prototype.zoomPercentSelect = function (args) {
        if (this.pdfViewerBase.pageCount > 0) {
            if (args.isInteracted) {
                if (args.itemData) {
                    // tslint:disable-next-line:no-any
                    var zoomText = args.itemData.percent;
                    this.zoomDropDownChange(zoomText);
                }
            }
            else {
                this.updateZoomPercentage(this.pdfViewer.magnificationModule.zoomFactor);
            }
        }
    };
    Toolbar.prototype.zoomDropDownChange = function (zoomText) {
        // tslint:disable-next-line:max-line-length
        if (zoomText !== this.pdfViewer.localeObj.getConstant('Fit Width') && zoomText !== this.pdfViewer.localeObj.getConstant('Fit Page') && zoomText !== this.pdfViewer.localeObj.getConstant('Automatic')) {
            this.pdfViewer.magnificationModule.isAutoZoom = false;
            this.pdfViewer.magnificationModule.zoomTo(parseFloat(zoomText));
            this.zoomDropDown.focusOut();
        }
        else if (zoomText === this.pdfViewer.localeObj.getConstant('Fit Width')) {
            this.pdfViewer.magnificationModule.isAutoZoom = false;
            this.pdfViewer.magnificationModule.fitToWidth();
            this.zoomDropDown.focusOut();
        }
        else if (zoomText === this.pdfViewer.localeObj.getConstant('Fit Page')) {
            this.pdfViewer.magnificationModule.fitToPage();
            this.zoomDropDown.focusOut();
        }
        else if (zoomText === this.pdfViewer.localeObj.getConstant('Automatic')) {
            this.pdfViewer.magnificationModule.isAutoZoom = true;
            this.pdfViewer.magnificationModule.fitToAuto();
            this.zoomDropDown.focusOut();
        }
    };
    /**
     * @private
     */
    Toolbar.prototype.updateZoomPercentage = function (zoomFactor) {
        // tslint:disable-next-line
        var currentPercent = parseInt((zoomFactor * 100).toString()) + '%';
        if (this.zoomDropDown.text === currentPercent) {
            this.zoomDropDown.element.value = currentPercent;
        }
        if (this.zoomDropDown.index === 9) {
            this.zoomDropDown.value = 2;
        }
        // tslint:disable-next-line
        this.zoomDropDown.text = currentPercent;
    };
    Toolbar.prototype.updateInteractionItems = function () {
        if (this.pdfViewer.textSelectionModule) {
            if (this.pdfViewer.enableTextSelection) {
                this.toolbar.enableItems(this.textSelectItem.parentElement, true);
            }
            else {
                this.toolbar.enableItems(this.textSelectItem.parentElement, false);
            }
        }
        else {
            this.toolbar.enableItems(this.textSelectItem.parentElement, false);
        }
        this.toolbar.enableItems(this.panItem.parentElement, true);
        if (this.pdfViewer.interactionMode === 'TextSelection') {
            this.selectItem(this.textSelectItem);
            this.deSelectItem(this.panItem);
        }
        else {
            this.selectItem(this.panItem);
            this.deSelectItem(this.textSelectItem);
            this.pdfViewerBase.initiatePanning();
        }
    };
    /**
     * @private
     */
    Toolbar.prototype.textSearchButtonHandler = function () {
        if (this.pdfViewer.textSearchModule && this.pdfViewerBase.pageCount > 0) {
            this.isTextSearchBoxDisplayed = !this.isTextSearchBoxDisplayed;
            this.pdfViewer.textSearchModule.showSearchBox(this.isTextSearchBoxDisplayed);
            if (this.isTextSearchBoxDisplayed) {
                this.selectItem(this.textSearchItem);
                // tslint:disable-next-line:max-line-length
                var searchInputElement = document.getElementById(this.pdfViewer.element.id + '_search_input');
                searchInputElement.select();
                searchInputElement.focus();
            }
            else {
                this.deSelectItem(this.textSearchItem);
            }
        }
    };
    Toolbar.prototype.selectItem = function (element) {
        element.classList.add('e-pv-select');
    };
    Toolbar.prototype.deSelectItem = function (element) {
        element.classList.remove('e-pv-select');
    };
    Toolbar.prototype.updateInteractionTools = function (isTextSelect) {
        if (isTextSelect) {
            this.selectItem(this.textSelectItem);
            this.deSelectItem(this.panItem);
        }
        else {
            this.selectItem(this.panItem);
            this.deSelectItem(this.textSelectItem);
        }
    };
    Toolbar.prototype.initialEnableItems = function () {
        if (this.pdfViewer.enableToolbar) {
            this.showToolbar(true);
        }
        else {
            this.showToolbar(false);
        }
        if (this.pdfViewer.enablePrint) {
            this.showPrintOption(true);
        }
        else {
            this.showPrintOption(false);
        }
        if (this.pdfViewer.enableDownload) {
            this.showDownloadOption(true);
        }
        else {
            this.showDownloadOption(false);
        }
        if (this.pdfViewer.enableTextSearch) {
            this.showSearchOption(true);
        }
        else {
            this.showSearchOption(false);
        }
    };
    Toolbar.prototype.showSeparator = function (toolbarItems) {
        if (!this.isOpenBtnVisible || (!this.isNavigationToolVisible && this.isOpenBtnVisible && this.isDownloadBtnVisible
            && this.isPrintBtnVisible) || (this.isOpenBtnVisible && toolbarItems.length === 1)) {
            this.applyHideToToolbar(false, 1, 1);
        }
        if ((!this.isNavigationToolVisible || !this.isMagnificationToolVisible) && !this.isOpenBtnVisible) {
            this.applyHideToToolbar(false, 8, 8);
        }
        if ((!this.isMagnificationToolVisible && !this.isSelectionBtnVisible && !this.isScrollingBtnVisible) ||
            (this.isMagnificationToolVisible && (!this.isSelectionBtnVisible && !this.isScrollingBtnVisible)) ||
            (!this.isMagnificationToolVisible && (this.isSelectionBtnVisible && this.isScrollingBtnVisible))) {
            this.applyHideToToolbar(false, 12, 12);
        }
    };
    Toolbar.prototype.applyToolbarSettings = function () {
        var toolbarSettingsItems = this.pdfViewer.toolbarSettings.toolbarItem;
        if (toolbarSettingsItems.indexOf('OpenOption') !== -1) {
            this.showOpenOption(true);
        }
        else {
            this.showOpenOption(false);
        }
        if (toolbarSettingsItems.indexOf('PageNavigationTool') !== -1) {
            this.showPageNavigationTool(true);
        }
        else {
            this.showPageNavigationTool(false);
        }
        if (toolbarSettingsItems.indexOf('MagnificationTool') !== -1) {
            this.showMagnificationTool(true);
        }
        else {
            this.showMagnificationTool(false);
        }
        if (toolbarSettingsItems.indexOf('SelectionTool') !== -1) {
            this.showSelectionTool(true);
        }
        else {
            this.showSelectionTool(false);
        }
        if (toolbarSettingsItems.indexOf('PanTool') !== -1) {
            this.showScrollingTool(true);
        }
        else {
            this.showScrollingTool(false);
        }
        if (toolbarSettingsItems.indexOf('PrintOption') !== -1) {
            this.showPrintOption(true);
        }
        else {
            this.showPrintOption(false);
        }
        if (toolbarSettingsItems.indexOf('DownloadOption') !== -1) {
            this.showDownloadOption(true);
        }
        else {
            this.showDownloadOption(false);
        }
        if (toolbarSettingsItems.indexOf('SearchOption') !== -1) {
            this.showSearchOption(true);
        }
        else {
            this.showSearchOption(false);
        }
        this.showSeparator(toolbarSettingsItems);
    };
    /**
     * @private
     */
    Toolbar.prototype.getModuleName = function () {
        return 'Toolbar';
    };
    return Toolbar;
}());
export { Toolbar };
