// @syncfusion/ej2-angular-grids/src/grid/aggregate-columns.directive.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ViewContainerRef, ContentChildren, ContentChild } from '@angular/core';
import { ComplexBase, ArrayBase, setValue } from '@syncfusion/ej2-angular-base';
import { Template } from '@syncfusion/ej2-angular-base';
var input = ['columnName', 'customAggregate', 'field', 'footerTemplate', 'format', 'groupCaptionTemplate', 'groupFooterTemplate', 'type'];
var outputs = [];
/**
 * `e-aggregate->e-column` directive represent a aggregate column of the Angular Grid.
 * ```html
 * <ejs-grid [dataSource]='data' allowPaging='true' allowSorting='true'>
 *   <e-columns>
 *     <e-column field='ID' width='100'></e-column>
 *     <e-column field='name' headerText='Name' width='100'></e-column>
 *   </e-columns>
 *   <e-aggregates>
 *     <e-aggregate>
 *       <e-columns>
 *         <e-column field='ID' type='Min'></e-column>
 *       </e-columns>
 *      </e-aggregate>
 *    </e-aggregates>
 * </ejs-grid>
 * ```
 */
var AggregateColumnDirective = /** @class */ (function (_super) {
    __extends(AggregateColumnDirective, _super);
    function AggregateColumnDirective(viewContainerRef) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        setValue('currentInstance', _this, _this.viewContainerRef);
        _this.registerEvents(outputs);
        return _this;
    }
    AggregateColumnDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ejs-grid>e-aggregates>e-aggregate>e-columns>e-column',
                    inputs: input,
                    outputs: outputs,
                    queries: {}
                },] },
    ];
    /** @nocollapse */
    AggregateColumnDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, },
    ]; };
    AggregateColumnDirective.propDecorators = {
        'footerTemplate': [{ type: ContentChild, args: ['footerTemplate',] },],
        'groupFooterTemplate': [{ type: ContentChild, args: ['groupFooterTemplate',] },],
        'groupCaptionTemplate': [{ type: ContentChild, args: ['groupCaptionTemplate',] },],
    };
    __decorate([
        Template(),
        __metadata("design:type", Object)
    ], AggregateColumnDirective.prototype, "footerTemplate", void 0);
    __decorate([
        Template(),
        __metadata("design:type", Object)
    ], AggregateColumnDirective.prototype, "groupFooterTemplate", void 0);
    __decorate([
        Template(),
        __metadata("design:type", Object)
    ], AggregateColumnDirective.prototype, "groupCaptionTemplate", void 0);
    return AggregateColumnDirective;
}(ComplexBase));
export { AggregateColumnDirective };
/**
 * AggregateColumn Array Directive
 * @private
 */
var AggregateColumnsDirective = /** @class */ (function (_super) {
    __extends(AggregateColumnsDirective, _super);
    function AggregateColumnsDirective() {
        return _super.call(this, 'columns') || this;
    }
    AggregateColumnsDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'ejs-grid>e-aggregates>e-aggregate>e-columns',
                    queries: {
                        children: new ContentChildren(AggregateColumnDirective)
                    },
                },] },
    ];
    /** @nocollapse */
    AggregateColumnsDirective.ctorParameters = function () { return []; };
    return AggregateColumnsDirective;
}(ArrayBase));
export { AggregateColumnsDirective };
